RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc

```diff
@@ -0,0 +1,776 @@
+      int j = rhtools_.getLayerWithOffset(cluster.hitsAndFractions()[0].first) - 1;
+      if (j < eidNLayers_ && seenClusters[j] < eidNClusters_) {
+        // get the pointer to the first feature value for the current batch, layer and cluster
+        float *features = &input.tensor<float, inputDimension>()(i, j, seenClusters[j], 0);
+
+        // fill features
+        *(features++) = float(cluster.energy() / float(trackster.vertex_multiplicity(k)));
+        *(features++) = float(std::abs(cluster.eta()));
+        *(features) = float(cluster.phi());
+
+        // increment seen clusters
+        seenClusters[j]++;
+      }
+    }
+
+    // zero-fill features of empty clusters in each layer (6)
+    for (int j = 0; j < eidNLayers_; j++) {
+      for (int k = seenClusters[j]; k < eidNClusters_; k++) {
+        float *features = &input.tensor<float, inputDimension>()(i, j, k, 0);
+        for (int l = 0; l < eidNFeatures_; l++) {
+          *(features++) = 0.f;
+        }
+      }
+    }
+  }
+
+  // run the inference (7)
+  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
+
+  // store regressed energy per trackster (8)
+  if (!eidOutputNameEnergy_.empty()) {
+    // get the pointer to the energy tensor, dimension is batch x 1
+    float *energy = outputs[0].flat<float>().data();
+
+    for (const int &i : tracksterIndices) {
+      tracksters[i].setRegressedEnergy(*(energy++));
+    }
+  }
+
+  // store id probabilities per trackster (8)
+  if (!eidOutputNameId_.empty()) {
+    // get the pointer to the id probability tensor, dimension is batch x id_probabilities.size()
+    int probsIdx = !eidOutputNameEnergy_.empty();
+    float *probs = outputs[probsIdx].flat<float>().data();
+
+    for (const int &i : tracksterIndices) {
+      tracksters[i].setProbabilities(probs);
+      probs += tracksters[i].id_probabilities().size();
+    }
+  }
+}
+
+void TrackstersMergeProducerV3::assignTimeToCandidates(std::vector<TICLCandidate> &resultCandidates) const {
+  for (auto &cand : resultCandidates) {
+    if (cand.tracksters().size() > 1) {  // For single-trackster candidates the timing is already set
+      float time = 0.f;
+      float timeErr = 0.f;
+      for (const auto &tr : cand.tracksters()) {
+        if (tr->timeError() > 0) {
+          auto invTimeESq = pow(tr->timeError(), -2);
+          time += tr->time() * invTimeESq;
+          timeErr += invTimeESq;
+        }
+      }
+      if (timeErr > 0) {
+        timeErr = 1. / timeErr;
+
+        cand.setTime(time * timeErr);
+        cand.setTimeError(sqrt(timeErr));
+      }
+    }
+  }
+}
+
+void TrackstersMergeProducerV3::printTrackstersDebug(const std::vector<Trackster> &tracksters,
+                                                     const char *label) const {
+  if (!debug_)
+    return;
+
+  int counter = 0;
+  for (auto const &t : tracksters) {
+    LogDebug("TrackstersMergeProducerV3")
```

Comment:
LogDebug is only compiled if EDM_ML_DEBUG is defined at compile time, hence runtime checks can be omitted here.